### ISSUE TITLE
RFC: Completing the Migration to Clang/LLVM for RISC-V OpenTitan software builds

### ISSUE THREAD

**Issue by luismarques** [lowRISC/opentitan#2459]
As a follow-up to the previous proposal on migrating to Clang/LLVM for RISC-V OpenTitan software builds, we wrote a document that proposes completing this migration process. The document analyses the impact of the migration up to this date and details the next steps for completing the migration process. You can find the complete document here: [RFC: Completing the Migration to Clang/LLVM for RISC-V OpenTitan software builds](https://docs.google.com/document/d/11b4OC-pAdyw_LV8tsS3CYWqhsxKEkxhbAii52YbCX8o/edit).

**Comment by asb** [comment]
Thanks for posting this Luis. Just to note that @sjgitty have discussed and taking into consideration the fact this is an incremental follow-up to a previous widely circulated and accepted proposal have tentatively queued this for discussion at the Technical Committee meeting next Wednesday. The TC will be prepared to consider approval if there are no major concerns raised in this next week. If it seems more discussion is needed, we will bump to the next meeting.

**Comment by asb** [comment]
I'm pleased to report the Technical Committee voted to approve this RFC yesterday. We took into account the wide positive reception by software contributors to the project (either directly in response to this RFC or the earlier proposal) and the fact conversations had been had with the DV team. The TC noted that this toolchain transition process had surfaced the fact some contributors had not tried to use the project-preferred toolchain build and appreciate that conversations are ongoing to help those contributors. Given the GCC toolchain is not being deprecated, this doesn't block the proposed timeline.

**Comment by lenary** [comment]
Yesterday, in #2503, I merged the change that makes clang fail the build if it finds any errors (the next step in the document).

In order to address some issues we were seeing with the toolchain I have merged the update in #2557. This new version of the toolchain (it is built slightly differently, the llvm and gcc versions have not changed) should allow us to test this new toolchain in the private CI (we were seeing issues before), which is what #2524 is testing out. We're not going to email around this update as we don't see it as necessary for developers - if they can run clang without problems, then they don't need to do it. We should document the expectation that developers who run into toolchain issues should ensure their toolchain is up to date (which we don't think is an onerous requirement). Where we expect toolchain bugs and their fixes to impact most developers, we will send around requests for developers to update their toolchains.

--

There is one issue I need to get to the bottom of still, to do with the use of this toolchain by partner organisations. My understanding is that some are on old versions of linux that may have difficulty running the binary release of the toolchain. A little more coordination may be required, but I don't see this as preventing us switching CI and most developers to using clang, because we'll still be doing a GCC build in CI as well.
